patina_dxe_core/src/cpu/hw_interrupt_protocol.rs

```diff
@@ -418,38 +422,46 @@ impl InterruptHandler for HwInterruptProtocolHandler {
 impl HwInterruptProtocolHandler {
     pub fn new(handlers: Vec<Option<HwInterruptHandler>>, aarch64_int: AArch64InterruptInitializer) -> Self {
         Self {
-            handlers: TplMutex::new(efi::TPL_HIGH_LEVEL, handlers, "Hardware Interrupt Lock"),
+            handlers: handlers.into_iter().map(RwLock::new).collect(),
             aarch64_int: TplMutex::new(efi::TPL_HIGH_LEVEL, aarch64_int, "AArch64 GIC Lock"),
         }
     }
 
     /// Internal implementation of interrupt related functions.
     pub fn register_interrupt_source(&self, interrupt_source: usize, handler: HwInterruptHandler) -> efi::Status {
-        if interrupt_source >= self.handlers.lock().len() {
+        if interrupt_source >= self.handlers.len() {
             return efi::Status::INVALID_PARAMETER;
         }
 
         let m_handler = handler as *const c_void;
 
         // If the handler is a null pointer, return invalid parameter
-        if m_handler.is_null() & self.handlers.lock()[interrupt_source].is_none() {
-            return efi::Status::INVALID_PARAMETER;
-        }
+        if let Some(mut rw_handler) = self.handlers[interrupt_source as usize].try_write() {
+            if m_handler.is_null() && (*rw_handler).is_none() {
+                return efi::Status::INVALID_PARAMETER;
+            }
 
-        if !m_handler.is_null() & self.handlers.lock()[interrupt_source].is_some() {
-            return efi::Status::ALREADY_STARTED;
-        }
+            if !m_handler.is_null() && (*rw_handler).is_some() {
+                return efi::Status::ALREADY_STARTED;
+            }
 
-        // If the interrupt handler is unregistered then disable the interrupt
-        let result = if m_handler.is_null() {
-            self.handlers.lock()[interrupt_source as usize] = None;
-            self.aarch64_int.lock().disable_interrupt_source(interrupt_source as u64)
+            // If the interrupt handler is unregistered then disable the interrupt
+            if m_handler.is_null() {
+                *rw_handler = None;
+                if let Err(err) = self.aarch64_int.lock().disable_interrupt_source(interrupt_source as u64) {
+                    return err.into();
+                }
+            } else {
+                *rw_handler = Some(handler);
+                if let Err(err) = self.aarch64_int.lock().enable_interrupt_source(interrupt_source as u64) {
```

Comment:
you might want to drop the lock before you enable the interrupt source, otherwise an interrupt could immediately occur and try and get the handler and panic on the try_read.